### PR TITLE
prevent loading of illegal modules in the browser

### LINKED ISSUE #6700: Support for tools like telefunc
### Describe the bug

[Telefunc](https://telefunc.com/) allows simple frontend to backend calls, avoiding direct call to `fetch`.

```ts
// The `CreateTodo.telefunc.js` file is not loaded:
// Telefunc transforms `*.telefunc.js` imports into a
// thin HTTP client.
import { onNewTodo } from './CreateTodo.telefunc.js'

...

await onNewTodo({ text })
```

Using telefunc + sveltekit can cause the following issues:
- `Cannot import $env/static/private into client-side code`
- `Cannot import $lib/server/xyz.telefunc.ts into client-side code`

Those errors happen at build time and in dev mode, but in dev mode, switching between pages can make it work again. See reproduction repo

### Reproduction

https://github.com/magne4000/sveltekit-telefunc-repro

`pnpm i && pnpm run dev`

The 3 links at the top have the 3 following cases:
- working
- failing because of $env/static/private import
- failing because of $lib/server

### Logs

```shell
[vite-plugin-svelte-kit] Cannot import $lib/server/test3.telefunc.ts into client-side code:
- .svelte-kit/generated/nodes/4.js
  - src/routes/failing/server-folder/+page.svelte
    - $lib/server/test3.telefunc.ts
error during build:
Error: Cannot import $lib/server/test3.telefunc.ts into client-side code:
- .svelte-kit/generated/nodes/4.js
  - src/routes/failing/server-folder/+page.svelte
    - $lib/server/test3.telefunc.ts
    at IllegalModuleGuard.assert_legal (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/utils.js:47:11)
    at IllegalModuleGuard.assert_legal (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/utils.js:49:9)
    at prevent_illegal_rollup_imports (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/utils.js:424:8)
    at file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/index.js:357:7
    at Array.forEach (<anonymous>)
    at Object.handler (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/index.js:349:25)
    at file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/rollup@2.78.1/node_modules/rollup/dist/es/shared/rollup.js:22710:40
    at async PluginDriver.hookParallel (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/rollup@2.78.1/node_modules/rollup/dist/es/shared/rollup.js:22632:17)
    at async file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/rollup@2.78.1/node_modules/rollup/dist/es/shared/rollup.js:23763:13
    at async catchUnfinishedHookActions (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/rollup@2.78.1/node_modules/rollup/dist/es/shared/rollup.js:23088:20)
```

```shell
Cannot import $env/static/private into client-side code:
- src/routes/failing/private-env/+page.svelte
  - $lib/test2.telefunc.ts
    - $env/static/private
Error: Cannot import $env/static/private into client-side code:
- src/routes/failing/private-env/+page.svelte
  - $lib/test2.telefunc.ts
    - $env/static/private
    at IllegalModuleGuard.assert_legal (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/utils.js:47:11)
    at IllegalModuleGuard.assert_legal (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/utils.js:49:9)
    at prevent_illegal_vite_imports (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/utils.js:437:8)
    at Object.result.component (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:93:9)
    at async Promise.all (index 1)
    at async render_response (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/runtime/server/page/render.js:82:16)
    at async render_page (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/runtime/server/page/index.js:276:10)
    at async resolve (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/runtime/server/index.js:231:17)
    at async respond (file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/runtime/server/index.js:301:20)
    at async file:///home/magne/workspace/sveltekit-telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.480_svelte@3.50.1+vite@3.1.0/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:406:22
```

### System Info

```shell
System:
    OS: Linux 5.19 undefined
    CPU: (16) x64 AMD Ryzen 7 5800X 8-Core Processor
    Memory: 23.40 GB / 31.26 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 16.15.0 - ~/.volta/tools/image/node/16.15.0/bin/node
    Yarn: 1.22.19 - ~/.volta/tools/image/node/16.15.0/bin/yarn
    npm: 8.10.0 - ~/.volta/tools/image/npm/8.10.0/bin/npm
  Browsers:
    Firefox: 104.0.2
  npmPackages:
    @sveltejs/adapter-auto: next => 1.0.0-next.72
    @sveltejs/kit: next => 1.0.0-next.480
    svelte: ^3.44.0 => 3.50.1
    vite: ^3.1.0 => 3.1.0
```

### Severity

serious, but I can work around it

### Additional Information

The `.telefunc.` files are untouched on server side, but are obviously transformed when used client side. There seems to be a mixup in the dependencies check.
I don't know how difficult of a fix this is, but a hacky workaround could be to exclude `.telefunc.` files from the check :shrug:.

### PR DESCRIPTION
**Pull request by Rich-Harris**
Closes #6700. Closes #7288. Right now, we prevent people from importing illegal modules (`$env/[type]/private`, `$lib/server/**` and `**/*.server.{ext}`) by analysing the module graph during SSR. This is a flawed approach for a variety of reasons:

* It doesn't account for systems like telefunc, which transform modules differently between SSR and non-SSR modes such that the dangerous `import` declarations are removed
* it means you only get the error when you hit a page directly, not when you soft-navigate to a page. (I was surprised to realise that this morning!)
* it applies even to pages that have `csr = false`

A much simpler approach is to simply prevent the modules in question from being imported in the browser at dev time. We can do this by adding a guard in a `load` hook that throws if the relevant conditions are met. (At build time we keep the existing graph analysis.)

There's a couple of minor drawbacks: we don't get the import chain, just the module ID, and the error overlay is dismissable. I think these trade-offs are well worth it.

A nice bonus is that we can get rid of the code for traversing the Vite module graph. Since we now only need to worry about the Rollup module graph, this also means we can get rid of a layer of indirection (sorry @tcc-sejohnson, it was good code! but no longer needed). That part is a WIP.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by tcc-sejohnson** [comment]
As long as we're retaining the same graph analysis in build as we were before I see no problem with this -- I wanted to do this when we originally implemented the feature, but just couldn't figure out how. Please ping me when this is ready-er for merge -- happy to review 😁

**Comment by Rich-Harris** [comment]
Ok I think this is ready now, marking it as such and crossing my fingers that the tests pass. One wrinkle: if the Vite error overlay appears before hydration, it will get nuked if `%sveltekit.body%` isn't wrapped in an element below the `<body>`, so I had to make that change to the test app.

I almost wonder if we should enforce that, since it's a bit of a footgun otherwise.

@tcc-sejohnson would be grateful if you do get a chance to review! The build stuff should be basically identical

File: packages/kit/src/exports/vite/graph_analysis/index.spec.js [hunk 1]
```diff
@@ -1,191 +0,0 @@
-import { describe } from '../../../utils/unit_test.js';
```
Comment (tcc-sejohnson) [inline packages/kit/src/exports/vite/graph_analysis/index.spec.js:1]: I was excited about being able to simplify the graph analysis, but this kinda killed all of my excitement. It seems like a step backwards -- we're removing our tests for making sure that, given an illegally-named module, we're going to throw with our graph analysis. Getting the same level of coverage with integration tests is going to be tricky, and it's also going to make it harder in the future -- if someone discovers an edge case we're missing, we suddenly have to find some contrived way to test it through integration tests rather than just adding an illegal chain here.

File: packages/kit/src/exports/vite/graph_analysis/index.js [hunk 2]
```diff
@@ -1,277 +1,81 @@
...
-		this.#seen.add(this.id);
-		for (const child of this.#module_info.importedModules) {
-			if (!this.#module_types.has(path.extname(this.id))) {
-				continue;
-			}
-			yield ViteImportGraph.#new_internal(this.#module_types, child, this.#seen);
+	return {
+		/** @param {string} id */
+		check: (id) => {
+			follow(id, []);
```
Comment (tcc-sejohnson) [inline packages/kit/src/exports/vite/graph_analysis/index.js:70]: default is already empty array

```suggestion
			follow(id);
```

**Review by tcc-sejohnson — CHANGES_REQUESTED** [review]

**Comment by tcc-sejohnson** [comment]
> One wrinkle: if the Vite error overlay appears before hydration, it will get nuked if %sveltekit.body% isn't wrapped in an element below the <body>, so I had to make that change to the test app.

Anecdotally, I've seen a good number of people in the Discord that remove the `div` wrapper for "easier" styling, so... might be a good idea to try to find a way.

File: packages/kit/src/exports/vite/graph_analysis/index.spec.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/graph_analysis/index.spec.js:1]: yeah, this was me being lazy — i couldn't really reuse much of the existing test code, so I cheated and deleted the file. reinstated tests, and added checks that the messages are being generated correctly

**Comment by Rich-Harris** [comment]
alright, things finally seem to be working again. @tcc-sejohnson if you have the bandwidth to take another look since you requested changes, that would be great — looking forward to getting this one merged 🙏

**Comment by lukaszpolowczyk** [comment]
> Anecdotally, I've seen a good number of people in the Discord that remove the div wrapper for "easier" styling, so... might be a good idea to try to find a way.

@tcc-sejohnson I in one project added to this `div` a `style="display: contents;"`, and it was enough for me.

I don't know if that can cause a problem, but I haven't noticed any.

File: packages/kit/src/exports/vite/graph_analysis/index.js [hunk 3]
```diff
@@ -1,277 +1,107 @@
...
-	 * @returns {boolean}
+	 * @param {string} id
+	 * @param {Array<{ id: string, dynamic: boolean }>} chain
 	 */
-	#is_kit_illegal(module_id) {
-		return ILLEGAL_IMPORTS.has(module_id);
-	}
+	function follow(id, chain) {
+		if (seen.has(id)) return;
+		seen.add(id);
```
Comment (tcc-sejohnson) [inline packages/kit/src/exports/vite/graph_analysis/index.js:44]: nothing is wrong here; I just wanted you to see what the heck GitHub thinks these lines correlate to in the original diff

**Review by tcc-sejohnson — APPROVED** [review]
LGTM! This should be a great improvement for the dev-time experience. Thank you!

**Comment by tcc-sejohnson** [comment]
> > Anecdotally, I've seen a good number of people in the Discord that remove the div wrapper for "easier" styling, so... might be a good idea to try to find a way.
> 
> @tcc-sejohnson I in one project added to this `div` a `style="display: contents;"`, and it was enough for me.
> 
> I don't know if that can cause a problem, but I haven't noticed any.

That won't work for everyone. To be clear, I don't think there's really any logical reason to remove the wrapper div -- it's easy enough to set it up however you want to -- but some people do it and we don't want to break their apps if possible 😁

**Comment by stephane-vanraes** [comment]
> That won't work for everyone. To be clear, I don't think there's really any logical reason to remove the wrapper div -- it's easy enough to set it up however you want to -- but some people do it and we don't want to break their apps if possible 😁

I am one of those removing the wrapper div, so maybe shedding some insights on that 🙂

The simple reason is that this div servers no purpose whatsoever in the _final product_, and it makes styling harder to a degree (you always have to remember that the body of your page is not `body` for the css).

As far as I understand this div is a crutch for shortcoming of a tool used in _development_ only and imho that is enough reason to try to find a way to make things work without, because utilities meant for development have no place in a production build.

**Comment by Rich-Harris** [comment]
> besides potentially stopping hydration from removing extensions there injected markup

this is the key — we've had multiple reports of exactly this sort of thing happening. it's very much _not_ about preventing dev-time issues — it's the sort of bug that's likely to manifest in deeply confusing ways in production and that you'll be lucky to even find out about. we added a note to the docs...

image

...but i'm not sure that's enough, because people will always look at it and assume they can safely remove it. my personal inclination is to require `%sveltekit.body%` to be inside an element other than `<body>`, but i'm going to open a discussion to see how large a trade-off that would be

**Comment by Rich-Harris** [comment]
https://github.com/sveltejs/kit/discussions/7585
